### PR TITLE
57 binlog encryption 2nd pr

### PR DESCRIPTION
**Pull request by robgolebiowski**

### CODE REVIEW

File: mysql-test/suite/binlog_encryption/t/encrypted_master.test
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/t/encrypted_master.test]: Seems to belong to rpl_encryption

File: mysql-test/suite/binlog_encryption/t/encrypted_master_lost_key.test
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/t/encrypted_master_lost_key.test]: Likewise

File: mysql-test/suite/binlog_encryption/t/encrypted_master_switch_to_unencrypted.cnf
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/t/encrypted_master_switch_to_unencrypted.cnf]: Orphaned file?

File: mysql-test/suite/binlog_encryption/t/encrypted_slave.test
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/t/encrypted_slave.test]: Seems to belong to rpl_encryption

File: mysql-test/suite/binlog_encryption/t/encryption_combo.test
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/t/encryption_combo.test]: Likewise

File: mysql-test/suite/rpl_encryption/t/rpl_binlog_errors-master.opt
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/rpl_binlog_errors-master.opt]: Redundant

File: mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt]: Redundant

File: mysql-test/suite/rpl_encryption/t/rpl_special_charset.opt
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/rpl_special_charset.opt]: Redundant

File: mysql-test/suite/rpl_encryption/t/rpl_split_statements.cnf
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/rpl_split_statements.cnf]: Orphaned

File: mysql-test/suite/rpl_encryption/t/rpl_transaction_before_commit_failure-master.opt
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/rpl_transaction_before_commit_failure-master.opt]: Merge to cnf

File: mysql-test/suite/sys_vars/r/all_vars.result
Comment (laurynas-biveinis) [inline mysql-test/suite/sys_vars/r/all_vars.result]: sys_vars.all_vars should not pass like that if sys_vars.encrypt_binlog_basic is now present

File: mysql-test/suite/sys_vars/t/encrypt_binlog_basic.test
Comment (laurynas-biveinis) [inline mysql-test/suite/sys_vars/t/encrypt_binlog_basic.test]: Static

File: include/my_crypt.h
Comment (laurynas-biveinis) [inline include/my_crypt.h:45]: How about MY_ATTRIBUTE((warn_unused_result)) for all non-void functions?

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: Leftover from a decription_event/fdle rename?

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: Likewise

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: Likewise

File: sql/log_event.h
Comment (laurynas-biveinis) [inline sql/log_event.h]: Please add a comment why we assume that corrupted event is encrypted

File: sql/rpl_binlog_sender.cc
Comment (laurynas-biveinis) [inline sql/rpl_binlog_sender.cc]: The previous PR question still stands

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: include/my_crypt.h
Comment (robgolebiowski) [reply include/my_crypt.h:45]: Added

File: mysql-test/suite/binlog_encryption/t/encrypted_master.test
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/t/encrypted_master.test]: Done

File: mysql-test/suite/binlog_encryption/t/encrypted_master_lost_key.test
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/t/encrypted_master_lost_key.test]: Done

File: mysql-test/suite/binlog_encryption/t/encrypted_master_switch_to_unencrypted.cnf
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/t/encrypted_master_switch_to_unencrypted.cnf]: I have restored this test file. Seems I had a problem with this test because I had not crc check implemented yet. Now it works.

File: mysql-test/suite/binlog_encryption/t/encrypted_slave.test
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/t/encrypted_slave.test]: Done

File: mysql-test/suite/binlog_encryption/t/encryption_combo.test
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/t/encryption_combo.test]: Done

File: mysql-test/suite/rpl_encryption/t/rpl_binlog_errors-master.opt
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/rpl_binlog_errors-master.opt]: Removed

File: mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt]: Done

File: mysql-test/suite/rpl_encryption/t/rpl_special_charset.opt
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/rpl_special_charset.opt]: Done

File: mysql-test/suite/rpl_encryption/t/rpl_split_statements.cnf
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/rpl_split_statements.cnf]: Removed

File: mysql-test/suite/rpl_encryption/t/rpl_transaction_before_commit_failure-master.opt
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/rpl_transaction_before_commit_failure-master.opt]: Removed cnf file and added slave opt file.

File: mysql-test/suite/sys_vars/r/all_vars.result
Comment (robgolebiowski) [reply mysql-test/suite/sys_vars/r/all_vars.result]: Right, done

File: mysql-test/suite/sys_vars/t/encrypt_binlog_basic.test
Comment (robgolebiowski) [reply mysql-test/suite/sys_vars/t/encrypt_binlog_basic.test]: Done

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Right, changed

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Changed

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Done

File: sql/log_event.h
Comment (robgolebiowski) [reply sql/log_event.h]: Added

File: sql/rpl_binlog_sender.cc
Comment (robgolebiowski) [reply sql/rpl_binlog_sender.cc]: I have looked over the code and I cannot really prove that it can throw an exception. I saw quite a number of crashes when I was sending garbage data to slave, so just to be sure I put catch there. I will remove try catch, the new operator from Log_event doest not throw exceptions.

Comment (laurynas-biveinis) [reply sql/rpl_binlog_sender.cc]: Please do, (and please check if this is not compiled with -fno-exceptions in the first place so that nothrow new would be also redundant)

File: include/my_crypt.h
Comment (percona-ysorokin) [inline include/my_crypt.h]: unused define

File: mysys_ssl/my_crypt.cc
Comment (percona-ysorokin) [inline mysys_ssl/my_crypt.cc]: `my_aes_crypt_init()` should not leave `ctx` allocated / initialized if it returns error code

File: mysys_ssl/my_crypt.cc
Comment (percona-ysorokin) [inline mysys_ssl/my_crypt.cc:337]: this case should be handled inside `my_aes_crypt_init()` (see above)

File: sql/event_crypt.cc
Comment (percona-ysorokin) [inline sql/event_crypt.cc]: this case should be handled inside `my_aes_crypt_init()`

File: mysys_ssl/my_crypt.cc
Comment (percona-ysorokin) [inline mysys_ssl/my_crypt.cc:342]: why calling `my_aes_crypt_finish()` when `my_aes_crypt_update()` returns error code?

File: sql/event_crypt.cc
Comment (percona-ysorokin) [inline sql/event_crypt.cc:11]: Is `off` used as a source of randomness here? Does not look good from the proper algorithm usage POV.

File: sql/event_crypt.cc
Comment (percona-ysorokin) [inline sql/event_crypt.cc:43]: Likewise. File position as a source of randomness.

File: sql/binlog_crypt_data.h
Comment (percona-ysorokin) [inline sql/binlog_crypt_data.h]: `const uchar* iv` ?

File: sql/binlog_crypt_data.cc
Comment (percona-ysorokin) [inline sql/binlog_crypt_data.cc]: `const uchar* iv` ?

File: sql/binlog_crypt_data.cc
Comment (percona-ysorokin) [inline sql/binlog_crypt_data.cc:118]: Which `iv` did you mean here? `iv` input parameter or `this->iv`? Looks like a bug.

File: mysys_ssl/my_crypt.cc
Comment (percona-ysorokin) [inline mysys_ssl/my_crypt.cc]: Why not to store `oiv` here as just a pointer `const uchar* oiv`?
2 places where `MyEncryptionCTX::init()` is called with `ENCRYPTION_FLAG_NOPAD`:
1. in `encrypt_event()` it outlives the call.
2. in `Event_encrypter::init()` it can be bound to `crypto->iv`.

File: sql/event_crypt.cc
Comment (percona-ysorokin) [inline sql/event_crypt.cc]: formatting

File: mysql-test/suite/binlog_encryption/r/encrypted_master.result
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/r/encrypted_master.result]: Orphaned

File: mysql-test/suite/binlog_encryption/r/encrypted_slave.result
Comment (laurynas-biveinis) [inline mysql-test/suite/binlog_encryption/r/encrypted_slave.result]: Orphaned

File: mysql-test/suite/rpl_encryption/t/encrypted_master_switch_to_unencrypted.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/encrypted_master_switch_to_unencrypted.test]: master-slave.inc

File: mysql-test/suite/rpl_encryption/t/encryption_init_error.test
Comment (laurynas-biveinis) [inline mysql-test/suite/rpl_encryption/t/encryption_init_error.test]: Belongs to binlog_encryption

File: include/my_crypt.h
Comment (robgolebiowski) [reply include/my_crypt.h]: Removed

File: mysys_ssl/my_crypt.cc
Comment (robgolebiowski) [reply mysys_ssl/my_crypt.cc]: It is much safer this way, you do not have keep this memory allocated and it would be weird to use such function. Besides, iv in Event_encrypter::init() is a local variable that would have to be allocated on heap. It is small array and imho if possible it should be allocated on stack.

File: mysys_ssl/my_crypt.cc
Comment (robgolebiowski) [reply mysys_ssl/my_crypt.cc]: Done

File: mysys_ssl/my_crypt.cc
Comment (robgolebiowski) [reply mysys_ssl/my_crypt.cc:337]: Done

File: mysys_ssl/my_crypt.cc
Comment (robgolebiowski) [reply mysys_ssl/my_crypt.cc:342]: Because my_aes_crypt_finish is doing clean up that should done if ctx was initialized.

File: sql/binlog_crypt_data.cc
Comment (robgolebiowski) [reply sql/binlog_crypt_data.cc]: This is output parameter.

File: sql/binlog_crypt_data.cc
Comment (robgolebiowski) [reply sql/binlog_crypt_data.cc:118]: It's not a bug, it's input parameter.

File: sql/event_crypt.cc
Comment (robgolebiowski) [reply sql/event_crypt.cc:11]: It's correct. iv has to be unique in combination with a key. It does not have to be random, it does not even have to be secret, it is usually send in plain text. It's there to secure against ecb kind of attacks.

File: sql/binlog_crypt_data.h
Comment (robgolebiowski) [reply sql/binlog_crypt_data.h]: it's output parameter.

File: sql/event_crypt.cc
Comment (robgolebiowski) [reply sql/event_crypt.cc:43]: The same answer as above.

File: sql/event_crypt.cc
Comment (robgolebiowski) [reply sql/event_crypt.cc]: Done

File: sql/event_crypt.cc
Comment (robgolebiowski) [reply sql/event_crypt.cc]: Done

File: sql/rpl_binlog_sender.cc
Comment (robgolebiowski) [reply sql/rpl_binlog_sender.cc]: nothrow is redundant because new operator of class Log_event is overloaded and uses my_malloc to allocate memory.

File: mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt
Comment (laurynas-biveinis) [reply mysql-test/suite/rpl_encryption/t/rpl_mts_gtids_restart_slave_io_lost_trx-slave.opt]: Still there

File: mysys_ssl/CMakeLists.txt
Comment (laurynas-biveinis) [inline mysys_ssl/CMakeLists.txt:43]: Does incremental YaSSL configuration rebuild work, if only yassl.cc is touched?

File: sql/binlog.cc
Comment (laurynas-biveinis) [inline sql/binlog.cc]: Indentation

File: sql/binlog.cc
Comment (laurynas-biveinis) [inline sql/binlog.cc]: Indentation

File: sql/binlog_crypt_data.cc
Comment (laurynas-biveinis) [inline sql/binlog_crypt_data.cc]: These are not swap semantics - confusing, but I am not sure whether it's worth fixing it (and proper fix would probably be easiest by going back to assignment operator implementation)

File: sql/binlog_crypt_data.cc
Comment (laurynas-biveinis) [inline sql/binlog_crypt_data.cc]: true

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: true

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: true

File: sql/log_event.cc
Comment (laurynas-biveinis) [inline sql/log_event.cc]: bool? false?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Please also start a Jenkins run

File: mysql-test/suite/binlog_encryption/r/encrypted_master.result
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/r/encrypted_master.result]: Removed

File: mysql-test/suite/binlog_encryption/r/encrypted_slave.result
Comment (robgolebiowski) [reply mysql-test/suite/binlog_encryption/r/encrypted_slave.result]: Removed

File: mysql-test/suite/rpl_encryption/t/encrypted_master_switch_to_unencrypted.test
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/encrypted_master_switch_to_unencrypted.test]: Done

File: mysql-test/suite/rpl_encryption/t/encryption_init_error.test
Comment (robgolebiowski) [reply mysql-test/suite/rpl_encryption/t/encryption_init_error.test]: Moved

File: sql/binlog.cc
Comment (robgolebiowski) [reply sql/binlog.cc]: Done

File: sql/binlog_crypt_data.cc
Comment (robgolebiowski) [reply sql/binlog_crypt_data.cc]: Changed

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Done

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Done

File: sql/log_event.cc
Comment (robgolebiowski) [reply sql/log_event.cc]: Fixed

File: sql/binlog.cc
Comment (robgolebiowski) [reply sql/binlog.cc]: Done

File: sql/binlog_crypt_data.cc
Comment (robgolebiowski) [reply sql/binlog_crypt_data.cc]: Yes, I agree. Moving back this function just to assign operator would be best. Having implement this swap function in proper way requires real swaps that are not needed in our case. I will put it back into assignment operator.

File: mysys_ssl/CMakeLists.txt
Comment (robgolebiowski) [reply mysys_ssl/CMakeLists.txt:43]: Yes, it works. Change to yassl.cc equals change to my_crypt.cc as it includes whole yassl.cc file.

**Comment by robgolebiowski** [comment]
http://jenkins.percona.com/job/mysql-5.7-param/1320/

File: libbinlogevents/include/binlog_event.h
Comment (laurynas-biveinis) [inline libbinlogevents/include/binlog_event.h]: its ids

File: mysql-test/include/search_pattern_in_file.inc
Comment (laurynas-biveinis) [inline mysql-test/include/search_pattern_in_file.inc:58]: I don't like "changelog" comments like that: too easy not to update, do not add any value over git log/blame. But OK to keep since it's already here

File: mysys_ssl/my_crypt.cc
Comment (laurynas-biveinis) [inline mysys_ssl/my_crypt.cc]: static

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
I have very minor comments only and do not need another review round after addressing them. Thus if/when other reviewers are OK:
- please squash, rebase on 5.7 trunk now that it has 5.7.20;
- retest Jenkins after 5.7.20 rebase for any surprises;
- please wait until this next release branches off before merging;
- please coordinate with docs and QA

**Comment by laurynas-biveinis** [comment]
Please also mention in the squashed commit message that you are fixing LP 1729321

**Comment by percona-ysorokin** [comment]
General recommendations before the merge:
1. Please, make sure you have everything OK in both OpenSSL and YaSSL builds.
1a. Double check platforms with OpenSSL 1.1
2. Run locally at least one `binlog_encrypt` MTR test suite under ASan/Valgrind.
3. Run full test suite with enforced binlog encryption
`./mysql-test/mtr --debug-server --mysqld=encrypt_binlog=ON ...`

File: libbinlogevents/include/binlog_event.h
Comment (robgolebiowski) [reply libbinlogevents/include/binlog_event.h]: Changed.

File: mysql-test/include/search_pattern_in_file.inc
Comment (robgolebiowski) [reply mysql-test/include/search_pattern_in_file.inc:58]: Noted.

File: mysys_ssl/my_crypt.cc
Comment (robgolebiowski) [reply mysys_ssl/my_crypt.cc]: Done.

**Comment by robgolebiowski** [comment]
Re-based on latest trunk
Jenkins run: http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1357/

**Pull request merged**
